torch/_inductor/codegen/cpp_prefix.h

```diff
@@ -60,60 +61,162 @@ struct IsVecType<at::vec::Vectorized<T>>: std::true_type {};
-    new_index,
+    a.m2 + b.m2 + delta * delta * a.weight * wb_over_w,
+    new_weight,
   };
   return result;
 }
 
 template <typename T>
-Welford<T> welford_combine(const Welford<T> &acc, T data, const WeightRecp<T>* w=nullptr) {
+Welford<T> welford_combine(const Welford<T> &acc, const T& data, const WeightRecp<T>* w=nullptr) {
   // Add a single data point
-  int64_t index = acc.index + 1;
+  int64_t new_index = acc.index + 1;
+  auto new_weight = acc.weight + T(1);
   auto delta = data - acc.mean;
   T new_mean;
   if constexpr (!IsVecType<T>::value) {
-    new_mean = acc.mean + delta / T(index);
+    new_mean = acc.mean + delta / new_weight;
   } else {
+    // use new_index to fecth 1 / new_weight to avoid divisions
     new_mean = acc.mean +
-      ((w == nullptr || acc.index >= w->weight_recps.size())
-            ? delta / T(index)
-            : delta * T(w->weight_recps[acc.index]));
+      ((w == nullptr || new_index <= 0 || new_index >= w->weight_recps.size())
+            ? delta / new_weight
+            : delta * w->weight_recps[new_index]);
   }
   auto new_delta = data - new_mean;
   auto result = Welford<T>{
     new_mean,
     acc.m2 + delta * new_delta,
-    index
+    new_weight,
+    new_index
   };
   return result;
 }
 
+#if INDUCTOR_USE_VECTOR_TYPES()
+template <typename T>
+Welford<T> welford_combine(const Welford<T>& a, const Welford<T>& b, const int64_t tail_size) {
+  auto delta = b.mean - a.mean;
+  auto new_weight = a.weight + b.weight;
+  auto wb_over_w = b.weight / new_weight;
+  // Guard against division by zero
+  wb_over_w = T::blendv(wb_over_w, T(0), new_weight == T(0));
+  auto new_mean = a.mean + delta * wb_over_w;
+  auto new_m2 = a.m2 + b.m2 + delta * delta * a.weight * wb_over_w;
+  return Welford<T>{
+    T::set(a.mean, new_mean, tail_size),
+    T::set(a.m2, new_m2, tail_size),
+    T::set(a.weight, new_weight, tail_size),
+  };
+}
+
+template <typename T>
+Welford<T> welford_combine(const Welford<T>& acc, const T& data, const int64_t tail_size, const WeightRecp<T>* w=nullptr) {
+  // Add a single data point
+  auto delta = data - acc.mean;
+  int64_t new_index = acc.index + 1;
+  auto new_weight = acc.weight + T(1);
+  T new_mean;
+  if constexpr (!IsVecType<T>::value) {
+    new_mean = acc.mean + delta / new_weight;
+  } else {
+    // use new_index to fecth 1 / new_weight to avoid divisions
+    new_mean = acc.mean +
+      ((w == nullptr || new_index <= 0 || new_index >= w->weight_recps.size())
+            ? delta / new_weight
+            : delta * w->weight_recps[new_index]);
+  }
+  auto new_delta = data - new_mean;
+  auto new_m2 = acc.m2 + delta * new_delta;
+  return Welford<T>{
+    T::set(acc.mean, new_mean, tail_size),
+    T::set(acc.m2, new_m2, tail_size),
+    T::set(acc.weight, new_weight, tail_size),
+    new_index
+  };
+}
+
+template <typename T>
+T reduce(const T& a, const T& b, const std::string& reduction_type) {
+  if (reduction_type == "max") {
```

Comment:
This essentially moves the compile time checks to runtime. I don't think it is the right thing to do.

Reply:
I have modified this part, please review it again, thanks!